<!DOCTYPE html>
<html lang="" xml:lang="">
  <head>
    <title>04_01_bring_your_own_data.knit</title>
    <meta charset="utf-8" />
    <meta name="author" content="Selina Baldauf" />
    <meta name="date" content="2021-08-01" />
    <script src="04_01_bring_your_own_data_files/header-attrs-2.20/header-attrs.js"></script>
    <link href="04_01_bring_your_own_data_files/remark-css-0.0.1/default.css" rel="stylesheet" />
    <link rel="stylesheet" href="css/new_slides.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">

# Organization: Schedule

**Schedule of today**

- **Now - 14** (or 14.30 if you are enthusiastic still): Work on the data set(s)
  
  - Take break(s) as best fits your needs
  
- **14 (14.30) - 15**: Short feedback round

  - What did you find out about your data set? Plots, summaries, ...
  - Which methods did you use?
  - Did you learn something new?
  - Was there something you struggled with?
  - ...
  
- **15-16**: Feedback, conclusion

---
# New methods: Correlation plots and PCAS

- [`corrplot` package](https://taiyun.github.io/corrplot/) for correlation plots
- [Tutorial for PCAs in R](http://www.sthda.com/english/articles/31-principal-component-methods-in-r-practical-guide/118-principal-component-analysis-in-r-prcomp-vs-princomp/), [PCA tutorial for penguins](https://biodash.github.io/codeclub/s03e03_pca/)
- [`factoextra` package](https://rpkgs.datanovia.com/factoextra/index.html) for PCA visualization
- Correlations and PCAs do not work with `NA` values: use `tidyr::drop_na()` to remove all `NA` values from the data first
- These plots work for: penguins, wine, piecrab (?) dataset



![](04_01_bring_your_own_data_files/figure-html/unnamed-chunk-1-1.png)&lt;!-- --&gt;![](04_01_bring_your_own_data_files/figure-html/unnamed-chunk-1-2.png)&lt;!-- --&gt;




---
# Data set 1: What makes a good wine?

**Physicochemical properties of wine and quality judgements**


```
## 'data.frame':	1599 obs. of  12 variables:
##  $ fixed.acidity       : num  12.7 9.8 6.5 8.6 7.5 7.6 10.1 6.4 6.1 6.7 ...
##  $ volatile.acidity    : num  0.6 0.66 0.88 0.52 0.58 0.5 0.935 0.4 0.58 0.46 ...
##  $ citric.acid         : num  0.49 0.39 0.03 0.38 0.14 0.29 0.22 NA 0.23 0.24 ...
##  $ residual.sugar      : num  2.8 3.2 NA 1.5 2.2 2.3 3.4 1.6 2.5 1.7 ...
##  $ chlorides           : num  0.075 0.083 0.079 0.096 0.077 NA 0.105 0.066 0.044 0.077 ...
##  $ free.sulfur.dioxide : num  5 21 23 5 27 5 11 5 16 18 ...
##  $ total.sulfur.dioxide: num  NA 59 47 18 60 NA 86 12 70 34 ...
##  $ density             : num  0.999 0.999 0.996 NA 0.996 ...
##  $ pH                  : num  3.14 3.37 NA 3.2 3.28 3.32 3.43 3.34 3.46 3.39 ...
##  $ sulphates           : num  0.57 0.71 0.5 0.52 0.59 NA 0.64 NA NA 0.6 ...
##  $ alcohol             : num  11.4 11.5 11.2 9.4 9.8 11.5 11.3 9.2 12.5 10.6 ...
##  $ quality             : int  5 7 4 5 5 6 4 5 6 6 ...
```

Reference: P. Cortez, A. Cerdeira, F. Almeida, T. Matos and J. Reis.
Modeling wine preferences by data mining from physicochemical properties. In Decision Support Systems, Elsevier, 47(4):547-553, 2009.

---
# Data set 1: What makes a good wine?

.pull-left[

#### Ideas - know methods

- Plotting of wine quality against chemical properties
- Plotting of distribution of the different chemical properties
- Summary tables of the data using `dplyr`

#### Hints

- Transform the quality column to a factor before plotting: use `dplyr::mutate` and `as.factor()` to tranform the column
- Try the `janitor::clean_names()` function

]

.pull-right[
![](img/wine.jpg)
.footnote-right[.small[By Frederik Vandaele - originally posted to Flickr as Château Pétrus, CC BY 2.0, https://commons.wikimedia.org/w/index.php?curid=5145286]]]

---
# Data set 1: What makes a good wine?

#### Ideas - new methods

- **Correlation plots**: How are the different wine properties correlated with each other?
- **PCA**: How are the wine properties related to each other?

---
# Data set 2: Paralympic games from 1980-2016

|variable |class     |description |
|:--------|:---------|:-----------|
|gender   |character | Binary gender |
|event    |character | Event name |
|medal    |character | Medal type |
|athlete  |character | Athlete name (LAST NAME first name|
|abb      |character | Country abbreviation |
|country  |character | Country name |
|grp_id   |integer   | Group ID as a count within team sports |
|type     |character | Type of sport |
|year     |double    | year of games |
|guide    |character | Guide (for vision impaired athletes) |
|pilot    |character | Pilot (for vision impaired athletes) |

Source: [International Paralympic Committee](https://db.ipc-services.org/sdms/hira/web/index) (provided by [tidytuesday](https://github.com/rfordatascience/tidytuesday/edit/master/data/2021/2021-08-03/readme.md))

---
# Data set 2: Paralympic games from 1980-2016

Get the data:


```r
athletes &lt;- readr::read_csv('https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2021/2021-08-03/athletes.csv')
```

#### Ideas - know methods

- Use dplyr to
  - Create summaries of medal counts for different groups
- Did the ratio of men/women winning medals change over time?
- Which countries were the most successful ones? Does this differ between sports type?
  - Which types of sports accumulated the most medals?
- Make plots such as:
  - Age distribution of athletes winning gold, silver and bronze
  - Compare the total number of medals over the years between winter and summer olympics

---
# Data set 2: Paralympic games from 1980-2016

#### Hints

- To reduce complexity of the data, first filter only the athletes that won a medal (`!is.na(medal)`)

---
# Data set 3: Crab data set

.pull-left[

**Atlantic marsh fiddler crab (*Minuca pugnax*)**

- Crab from Florida is expanding northward due to ocean warming
- Data on 13 marshes across a range of latitude in the USA
- Recording of the size of of the crab
- Rather small and good to handle


]

.pull-right[

.center[![](img/day4/lter.jpg)]

.small[Source: [Johnson, D. 2019.](https://doi.org/10.6073/pasta/4c27d2e778d3325d3830a5142e3839bb) Fiddler crab body size in salt marshes from Florida to Massachusetts, USA at PIE and VCR LTER and NOAA NERR sites during summer 2016. ver 1. Environmental Data Initiative.]

]


.footnote-right[.small[Image under CC BY-SA 4.0]]

---
# Data set 3: Crab data set


```
## # A tibble: 6 × 9
##   date       latitude site   size air_temp air_temp_sd water_temp water_…¹ name 
##   &lt;date&gt;        &lt;dbl&gt; &lt;chr&gt; &lt;dbl&gt;    &lt;dbl&gt;       &lt;dbl&gt;      &lt;dbl&gt;    &lt;dbl&gt; &lt;chr&gt;
## 1 2016-07-24       30 GTM    12.4     21.8        6.39       24.5     6.12 Guan…
## 2 2016-07-24       30 GTM    14.2     21.8        6.39       24.5     6.12 Guan…
## 3 2016-07-24       30 GTM    14.5     21.8        6.39       24.5     6.12 Guan…
## 4 2016-07-24       30 GTM    12.9     21.8        6.39       24.5     6.12 Guan…
## 5 2016-07-24       30 GTM    12.4     21.8        6.39       24.5     6.12 Guan…
## 6 2016-07-24       30 GTM    13.0     21.8        6.39       24.5     6.12 Guan…
## # … with abbreviated variable name ¹​water_temp_sd
```

.small[Source: [Johnson, D. 2019.](https://doi.org/10.6073/pasta/4c27d2e778d3325d3830a5142e3839bb) Fiddler crab body size in salt marshes from Florida to Massachusetts, USA at PIE and VCR LTER and NOAA NERR sites during summer 2016. ver 1. Environmental Data Initiative.]


.footnote-right[.small[Image by LTER under CC BY-SA 4.0]]

---
# Data set 3: Crab data set

.pull-left[

**Ideas - known methods**

- Explore Bergmann's rule (organisms are large in higher latitudes)
- t-tests to compare size between locations
- Plot relationship between latitude and size 
- Plot distributions of variables

]

.pull-right[

.center[![](img/day4/lter.jpg)]

.small[Source: [Johnson, D. 2019.](https://doi.org/10.6073/pasta/4c27d2e778d3325d3830a5142e3839bb) Fiddler crab body size in salt marshes from Florida to Massachusetts, USA at PIE and VCR LTER and NOAA NERR sites during summer 2016. ver 1. Environmental Data Initiative.]

]

.footnote-right[.small[Image under CC BY-SA 4.0]]

---
# Data set 4: Ice cover and temperature

.pull-left[
**Temperature and ice duration on lakes since 19th century**

- 2 data sets with measurements of
  - ice start, end and duration on 2 lakes in Wisconsin
  - daily air temperature since 1870
- Explore the effect of climate change on ice cover
]
.pull-right[

![](img/day4/ice.jpg)

]

.footnote-right[.small[Image by LTER under CC BY-SA 4.0]]

.small[

Source ice data: [Magnuson, J.J., S.R. Carpenter, and E.H. Stanley. 2021.](https://doi.org/10.6073/pasta/ab31f2489ee436beb73fc8f1d0213d97) North Temperate Lakes LTER: Ice Duration - Madison Lakes Area 1853 - current ver 35. Environmental Data Initiative. 

Source temperature data: [Anderson, L. and D. Robertson. 2020.](https://doi.org/10.6073/pasta/e3ff85971d817e9898bb8a83fb4c3a8b) Madison Wisconsin Daily Meteorological Data 1869 - current ver 32. Environmental Data Initiative.
]

---
# Data set 4: Ice cover and temperature

Ice data:


```
## # A tibble: 6 × 5
##   lakeid       ice_on     ice_off    ice_duration  year
##   &lt;fct&gt;        &lt;date&gt;     &lt;date&gt;            &lt;dbl&gt; &lt;dbl&gt;
## 1 Lake Mendota NA         1853-04-05           NA  1852
## 2 Lake Mendota 1853-12-27 NA                   NA  1853
## 3 Lake Mendota 1855-12-18 1856-04-14          118  1855
## 4 Lake Mendota 1856-12-06 1857-05-06          151  1856
## 5 Lake Mendota 1857-11-25 1858-03-26          121  1857
## 6 Lake Mendota 1858-12-08 1859-03-14           96  1858
```


Temperature data:


```
## # A tibble: 6 × 3
##   sampledate  year ave_air_temp_adjusted
##   &lt;date&gt;     &lt;dbl&gt;                 &lt;dbl&gt;
## 1 1870-06-05  1870                  20  
## 2 1870-06-06  1870                  18.3
## 3 1870-06-07  1870                  17.5
## 4 1870-06-09  1870                  13.3
## 5 1870-06-10  1870                  13.9
## 6 1870-06-11  1870                  15
```

---
# Data set 4: Ice cover and temperature

.pull-left[
**Ideas - known methods**

- How did ice cover duration change over the years?
- How did air temperature change over the years?
  - Summarize mean annual temperature or mean temperature in winter
- How do ice duration on the lakes correlate with temperature (e.g. with mean winter 
  temperature)
  
]
.pull-right[

![](img/day4/ice.jpg)]

.footnote-right[.small[Image by LTER under CC BY-SA 4.0]]

#### Hints

- For some plots it might make sense to summarize annual means first
- Use `left_join` to combined the tables with annual mean temperature and ice duration

---
# Data set 5: RNAseq data

- Data from [FU et al. 2015, *Nature Cell Biology*](https://www.nature.com/articles/ncb3117)
- Data found via [Tutorial on heat maps](https://training.galaxyproject.org/training-material/topics/transcriptomics/tutorials/rna-seq-viz-with-heatmap2/tutorial.html) using this data

- 3 csv files:
  - `heatmap_genes.csv`: A list of the names of interesting genes to look at (Genes used in Figure 6b in paper)
  - `DE_results.csv`: Gene expression in luminal cells in pregnant versus lactating mice
    - logFC, AveExpr, t, p-value
    - Also contains non-significantly expressed genes
  - `normalized_counts`: Normalized counts for genes for the different samples
  
---
# Data set 5: RNAseq data

.pull-left[

#### Ideas:

- Create a heatmap of the top 20 most significant genes (see plot in the tutorial)
- Create a heatmap of the interesting genes (see Fig. 6 in the paper)
- Crate a volcano plot of the data similar to the one [here](https://training.galaxyproject.org/training-material/topics/transcriptomics/tutorials/rna-seq-viz-with-volcanoplot/tutorial.html)
]

.pull-right[

![](img/day4/volcano.png)
]

---
# Data set 5: RNAseq data

.pull-left[

![](img/day4/heat_top20.png)
]

.pull-right[
![](img/day4/heat_top20_2.png)
]

---
# Data set 5: RNAseq data

#### Some tips:

Data cleaning:
  - Read in the data and then use the `janitor::clean_names` function to make the column headers nicer
  - Join `DE_results` and `normalized_counts` by their shared columns
  - Use `select` to remove columns you don't need for analysis to get a better overview
  - Filter only significant genes ([tutorial](https://training.galaxyproject.org/training-material/topics/transcriptomics/tutorials/rna-seq-viz-with-volcanoplot/tutorial.html)) defines them as `p_value &lt; 0.01 &amp; abs(logFC)` &gt; 0.58)
  
Data analysis:
  - Heatmap with ggplot or with `pheatmap::pheatmap()`
    - `pheatmap` takes a matrix as input (use `as_matrix` on tibble to transform)
  - scale the counts -&gt; have a look at the `scale` function
    - `pheatmap` can scale but with ggplot you have to scale before plotting

---
# Some general tips

- First make a plan:

  - What do you want to achieve and what are the steps?
  - Try to think in technical terms

--

- Start with something small, e.g. reading in the data and bringing it into the right format.

--

- Google

--

- If you get stuck, ask in the chat or stop by in General

--

- Have a look at the cheat sheets [(list of all cheat sheets)](https://www.rstudio.com/resources/cheatsheets/)
  - [ggplot2](https://raw.githubusercontent.com/rstudio/cheatsheets/master/data-visualization-2.1.pdf)
  - [dplyr](https://raw.githubusercontent.com/rstudio/cheatsheets/master/data-transformation.pdf)
  - [Data import cheat sheet](https://raw.githubusercontent.com/rstudio/cheatsheets/master/data-import.pdf): readr and tidyr cheat sheet

---
class: inverse, middle, center

# .large[Now You: Working on your own research data]

#### - Meet in your group (if you want)
#### - Start working on a data set
#### - Take breaks as you need and be back at 14:00
#### - Keep an eye on your group and the general chat

 
---
class: inverse, middle, center

# .large[Now You: Working on a data set from me]

#### - Think about the data set you would like to work on 
#### - Build groups
#### - Meet in your group
#### - Start working on a data set
#### - Take breaks as you need and be back at 14:00
#### - Keep an eye on your group and the general chat

---
# Sharing

In 1-2 mins:

- What was the highlight of your analysis?
  - Your favorite plot
  - Some cool code
  - A problem that you finally solved
  - Something new you learned

- What was difficult?

- If you want: Share a screenshot in the chat or share your screen

---
# Feedback

Please take 10 mins to complete the feedback survey for the Graduate center (don't use Internet Explorer)

.center[https://votingo.cedis.fu-berlin.de/PCNLP3]

---
# Feedback


- Any other feedback or comments from your side?

---
# Conclusion

![](img/day1/datascience_workflow_tidyverse.png)

We learned a lot of stuff!

---
# Conclusion

#### How to continue from here?

.pull-left[

- Learning by doing!
- Have a look at some [online ressoures](https://selinazitrone.github.io/intro-r-data-analysis/online_resources.html), I recommend the R for Data Science book by Hadley Wickham
- If you use Twitter: Follow some people that post R content regarding your interest
- If you like plotting: Consider participating in the [tidytuesday](https://github.com/rfordatascience/tidytuesday)
- [FU statistical consulting](http://www.stat.fu-berlin.de/en/dienstleistungen/index.html) for questions regarding statistical methods
]
- [R Consulting by me](https://www.bcp.fu-berlin.de/en/graduiertenzentrum/promovierende/beratung/index.html)
- Workshop on reproducible documents with Rmarkdown this summer
- Monthly R/programming topic

.pull-right[

![](img/AllisonHorst/r_rollercoaster.png)
]

.footnote-right[Artwork by [Allison Horst](https://twitter.com/allison_horst)]

---
class: inverse, middle, center

# .large[The End]
## Thanks a lot for participating!

![](img/AllisonHorst/heartyr.gif)

.footnote-right[Artwork by [Allison Horst](https://twitter.com/allison_horst)]
    </textarea>
<style data-target="print-only">@media screen {.remark-slide-container{display:block;}.remark-slide-scaler{box-shadow:none;}}</style>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script src="macros.js"></script>
<script>var slideshow = remark.create({
"highlightStyle": "github",
"highlightLines": true,
"ratio": "16:9",
"countIncrementalSlides": false
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function(d) {
  var s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})(document);

(function(d) {
  var el = d.getElementsByClassName("remark-slides-area");
  if (!el) return;
  var slide, slides = slideshow.getSlides(), els = el[0].children;
  for (var i = 1; i < slides.length; i++) {
    slide = slides[i];
    if (slide.properties.continued === "true" || slide.properties.count === "false") {
      els[i - 1].className += ' has-continuation';
    }
  }
  var s = d.createElement("style");
  s.type = "text/css"; s.innerHTML = "@media print { .has-continuation { display: none; } }";
  d.head.appendChild(s);
})(document);
// delete the temporary CSS (for displaying all slides initially) when the user
// starts to view slides
(function() {
  var deleted = false;
  slideshow.on('beforeShowSlide', function(slide) {
    if (deleted) return;
    var sheets = document.styleSheets, node;
    for (var i = 0; i < sheets.length; i++) {
      node = sheets[i].ownerNode;
      if (node.dataset["target"] !== "print-only") continue;
      node.parentNode.removeChild(node);
    }
    deleted = true;
  });
})();
// add `data-at-shortcutkeys` attribute to <body> to resolve conflicts with JAWS
// screen reader (see PR #262)
(function(d) {
  let res = {};
  d.querySelectorAll('.remark-help-content table tr').forEach(tr => {
    const t = tr.querySelector('td:nth-child(2)').innerText;
    tr.querySelectorAll('td:first-child .key').forEach(key => {
      const k = key.innerText;
      if (/^[a-z]$/.test(k)) res[k] = t;  // must be a single letter (key)
    });
  });
  d.body.setAttribute('data-at-shortcutkeys', JSON.stringify(res));
})(document);
(function() {
  "use strict"
  // Replace <script> tags in slides area to make them executable
  var scripts = document.querySelectorAll(
    '.remark-slides-area .remark-slide-container script'
  );
  if (!scripts.length) return;
  for (var i = 0; i < scripts.length; i++) {
    var s = document.createElement('script');
    var code = document.createTextNode(scripts[i].textContent);
    s.appendChild(code);
    var scriptAttrs = scripts[i].attributes;
    for (var j = 0; j < scriptAttrs.length; j++) {
      s.setAttribute(scriptAttrs[j].name, scriptAttrs[j].value);
    }
    scripts[i].parentElement.replaceChild(s, scripts[i]);
  }
})();
(function() {
  var links = document.getElementsByTagName('a');
  for (var i = 0; i < links.length; i++) {
    if (/^(https?:)?\/\//.test(links[i].getAttribute('href'))) {
      links[i].target = '_blank';
    }
  }
})();
// adds .remark-code-has-line-highlighted class to <pre> parent elements
// of code chunks containing highlighted lines with class .remark-code-line-highlighted
(function(d) {
  const hlines = d.querySelectorAll('.remark-code-line-highlighted');
  const preParents = [];
  const findPreParent = function(line, p = 0) {
    if (p > 1) return null; // traverse up no further than grandparent
    const el = line.parentElement;
    return el.tagName === "PRE" ? el : findPreParent(el, ++p);
  };

  for (let line of hlines) {
    let pre = findPreParent(line);
    if (pre && !preParents.includes(pre)) preParents.push(pre);
  }
  preParents.forEach(p => p.classList.add("remark-code-has-line-highlighted"));
})(document);</script>

<script>
slideshow._releaseMath = function(el) {
  var i, text, code, codes = el.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
};
slideshow._releaseMath(document);
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
